### ISSUE TITLE
DllImportGenerator crashes in IDE while editing GeneratedDllImport attribute

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#65008]
Repro:
Add a method like:
```C#
partial class Example
{
    [GeneratedDllImport("lib")]
    public static partial bool Func(bool arg);
}
```
Then start typing `, SetLastError=true` in the attribute, slowly enough that the generator kicks in mid-typing.  The generator will crash.

The problem is here:
https://github.com/dotnet/runtime/blob/a7027128ad6b0bd879b4cf2af6ef8686d725fcc9/src/libraries/System.Runtime.InteropServices/gen/DllImportGenerator/DllImportGenerator.cs#L388
While the attribute isn't fully formed, ConstructorArguments is Length 0, and accessing ConstructorArguments[0] produces an IndexOutOfRangeException.

cc: @dotnet/interop-contrib

**Comment by AaronRobinsonMSFT** [comment]
> Then start typing , SetLastError=true in the attribute, slowly enough that the generator kicks in mid-typing. The generator will crash.

Boo. This sort of issue makes things hard to test as well as validate because of the implicit UX component. Are there other areas this could occur?

**Comment by jkoritzinsky** [comment]
@AaronRobinsonMSFT we can test this by adding some "invalid" code snippets in our generator tests and making sure that we don't throw exceptions when we encounter invalid source that is close enough to provide some symbol information.

**Comment by AaronRobinsonMSFT** [comment]
> we can test this by adding some "invalid" code snippets in our generator tests and making sure that we don't throw exceptions when we encounter invalid source that is close enough to provide some symbol information.

@jkoritzinsky I like this idea a lot. I'd argue this is a concept that needs to be added to the Roslyn source generator "best-practices" guide. Since the source generator runs on potentially invalid code when in an IDE scenario it is paramount to have that testing in place to ensure a stable IDE experience.

/cc @chsienki

**Comment by AraHaan** [comment]
I agree, you can never be to careful what anyone like me could cook up that can break the entire IDE experience resulting in issues like this.

This is why even on my own personal generators, I unit test everything to 100%, and even then I unit test on things that may not be properly set up for it as well. In that case however my generator produces diagnostics (like analyzers) to inform the user about it, and then test to see if that is produced to a success. Perhaps the same if not similar thing can be done for this generator as well (perhaps something on the type to where if that thing checked is == 0 then report diagnostic and be done with it).

**Comment by AaronRobinsonMSFT** [comment]
Closed with #65344
